### PR TITLE
Upgrade tcmalloc-rs to 0.3

### PR DESCRIPTION
**Pull request by brson**
## What have you changed? (mandatory)

This is a follow up to https://github.com/tikv/tikv/pull/4370.

Upgrading to 0.3 links tcmalloc statically.

Note the [caveats about profiling and libunwind/libgcc on 64-bit linux][4] &mdash; they are not built by this crate, and tcmalloc will silently link to versions of both that can cause profiling deadlocks.

[4]: https://github.com/gperftools/gperftools/blob/master/INSTALL#L33

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

`cargo build --no-default-features --features=tcmalloc`

## Does this PR affect documentation (docs) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Refer to a related PR or issue link (optional)

Fixes https://github.com/tikv/tikv/issues/4471

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,5 +1,11 @@
 SHELL := /bin/bash
-ENABLE_FEATURES ?= default
+ENABLE_FEATURES ?=
+
+ifeq ($(TCMALLOC),1)
+ENABLE_FEATURES += tcmalloc
+else
+ENABLE_FEATURES += jemalloc
```
Comment (kennytm) [inline Makefile]: This means we can't choose the system allocator (disabling both `tcmalloc` & `jemalloc`) when using the makefile. Could we change the options to allow this?

**Comment by siddontang** [comment]
oh, can we link libunwind statically too?

File: Makefile (hunk 1 above)
Comment (siddontang) [reply Makefile]: +1

Comment (brson) [reply Makefile]: Done.

**Comment by brson** [comment]
@siddontang I think so, but unwinders are tricky so I'd like to do it as a followup. Here's an issue for it: https://github.com/tikv/tikv/issues/4520

**Comment by brson** [comment]
Comments addressed.

**Comment by siddontang** [comment]
@breeswish 

I think after this PR, you can update your `cargo profile` PR

**Pull request merged**